core/src/main/scala/chisel3/Layer.scala

```diff
@@ -51,6 +52,25 @@ object layer {
       case _          => s"${parent.fullName}.$name"
     }
 
+    /** The output directory of this layer.
+      *
+      * The output directory of a layer serves as a hint to the toolchain,
+      * specifying where files related to the layer (such as a bindfile, in
+      * verilog) should be output. If a layer has no output directory, then files
+      * related to this layer will be placed in the default output directory.
+      *
+      * Unless overridden, the outputDir's name matches the name of the layer,
+      * and is located under the parent layer's directory.
+      */
+    private[chisel3] def outputDir: Option[Path] = parent match {
+      case Layer.Root => Some(Paths.get(name))
+      case _ =>
+        parent.outputDir match {
+          case Some(dir) => Some(dir.resolve(name))
+          case None      => Some(Paths.get(name))
+        }
+    }
```

Comment:
While this works as `private[chisel3]`, it's not a great API for users as this isn't easily discoverable that this can be overridden. (None of these package private members show up in documentation and I'm not sure how LSPs will handle this.) There are two ways this could be made better:

1. Make this public. I'm not really in favor of this as I'm not big on APIs which are override based unless there is no other way.
2. Add a constructor `val` argument, like `convention` of type `Option[Path]`. This is also even better as a custom sum type as opposed to `Option`. Then keep a package private `final def` which will return a `Path` that represents the actual output directory that will be used.

WDYT?

Reply:
I agree that an argument is better than overriding virtual methods.